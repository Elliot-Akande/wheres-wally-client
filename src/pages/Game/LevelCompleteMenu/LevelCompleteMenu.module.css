.container {
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;

  display: flex;
  align-items: center;
  justify-content: center;

  background: #0000007c;
}

.menu {
  max-width: 250px;
  min-height: 300px;

  background: tomato;
}
