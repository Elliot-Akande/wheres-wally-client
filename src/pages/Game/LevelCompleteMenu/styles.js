import { keyframes, styled } from "styled-components";

const menu = keyframes`
  0% {
    transform: translateY(-200%) scale(0.4);
  }

  0%,
  25% {
    opacity: 0;
  }
`;

const container = keyframes`
  0% {
    background-color: transparent;
  }
`;

export const Wrapper = styled.div`
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;

  display: flex;
  align-items: center;
  justify-content: center;

  background: #0000007c;
  animation: ${container} 0.3s;
`;

export const Menu = styled.div`
  max-width: min-content;
  padding: 32px;

  display: flex;
  flex-direction: column;
  gap: 16px;

  background-color: var(--neutral-700);
  border-radius: 16px;
  animation: ${menu} 0.3s;
`;

export const Heading = styled.h1`
  margin: 0;
  font-size: 24px;
`;

export const Score = styled.p`
  margin: 0;
`;

export const Input = styled.input`
  width: 100%;
  padding: 4px 8px;
  margin-top: 4px;
  margin-bottom: 8px;

  color: var(--neutral-100);
  background-color: var(--neutral-500);
  border: 2px solid var(--neutral-400);
  border-radius: 8px;

  &:focus-visible {
    border-color: var(--accent);
    outline: none;
  }
`;

export const Button = styled.button`
  width: 100%;
  padding: 8px 16px;

  font-weight: bold;
  text-decoration: none;
  border: none;
  border-radius: 50px;
  color: var(--neutral-100);
  background: linear-gradient(180deg, var(--accent), var(--accent-dark));

  transition: 0.1s;

  &:hover {
    transform: scale(1.04);
  }
`;
