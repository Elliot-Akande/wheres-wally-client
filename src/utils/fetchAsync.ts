async function fetchAsync(uri, opts = {}) {
  const api = import.meta.env.VITE_API_URL;
  try {
    const response = await fetch(api + uri, opts);
    const jsonData = await response.json();
    if (!response.ok) {
      const err = new Error(jsonData);
      err.status = response.status;
      throw err;
    }

    return jsonData;
  } catch (err) {
    err.message = "An error occured when fetching data.";
    return err;
  }
}

export default fetchAsync;
