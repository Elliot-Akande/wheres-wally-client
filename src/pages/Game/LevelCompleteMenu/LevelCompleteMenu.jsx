import { useState } from "react";
import { useNavigate } from "react-router-dom";
import Leaderboard from "../../../components/Leaderboard/Leaderboard";
import useFetch from "../../../hooks/useFetch";
import fetchAsync from "../../../utils/fetchAsync";
import formatTime from "../../../utils/formatTime";
import styles from "./LevelCompleteMenu.module.css";

const LevelCompleteMenu = ({ levelNum, score, token }) => {
  const { data, loading, error } = useFetch(`/levels/${levelNum}/leaderboard`);
  const [updatedData, setUpdatedData] = useState(null);
  const [name, setName] = useState("Anon");
  const navigate = useNavigate();

  if (error) {
    return (
      <div className={styles.container}>
        <div className={styles.menu}>
          <h2>Level complete!</h2>
          <p>There was an error contacting the server.</p>
          <button type="button">Home</button>
        </div>
      </div>
    );
  }

  const handleSubmit = async (e) => {
    e.preventDefault();
    e.currentTarget.disabled = true;
    const response = await fetchAsync(`/levels/${levelNum}/leaderboard`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ name }),
    });

    if (!(response instanceof Error)) {
      setUpdatedData(response);
    }
  };

  const getMenuBody = () => {
    if (updatedData !== null) {
      return (
        <>
          <Leaderboard data={updatedData} loading={loading} />
          <button onClick={() => navigate("/")}>Finish</button>
        </>
      );
    }

    return (
      <>
        <Leaderboard data={data} loading={loading} />
        <p>Your Score - {formatTime(score)}</p>
        <form action="">
          <label htmlFor="name">Name</label>
          <input
            type="text"
            id="name"
            value={name}
            onChange={(e) => setName(e.target.value)}
          />
          <button type="submit" onClick={handleSubmit}>
            Submit
          </button>
        </form>
      </>
    );
  };

  return (
    <div className={styles.container}>
      <div className={styles.menu}>
        <h2>Level complete!</h2>
        {getMenuBody()}
      </div>
    </div>
  );
};

export default LevelCompleteMenu;
