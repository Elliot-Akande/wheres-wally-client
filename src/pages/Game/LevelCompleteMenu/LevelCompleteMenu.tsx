import PropTypes from "prop-types";
import { useState } from "react";
import { useNavigate } from "react-router-dom";
import Leaderboard from "../../../components/Leaderboard/Leaderboard";
import useFetch from "../../../hooks/useFetch";
import fetchAsync from "../../../utils/fetchAsync";
import formatTime from "../../../utils/formatTime";
import { Button, Input, Menu, Score, Wrapper, Heading } from "./styles";

const LevelCompleteMenu = ({ levelNum, score, token }) => {
  const { data, loading, error } = useFetch(`/levels/${levelNum}/leaderboard`);
  const [updatedData, setUpdatedData] = useState(null);
  const [name, setName] = useState("Anon");
  const navigate = useNavigate();

  if (error) {
    return (
      <Wrapper>
        <Menu>
          <h2>Level complete!</h2>
          <p>There was an error contacting the server.</p>
          <Button type="button" onClick={() => navigate("/")}>
            Home
          </Button>
        </Menu>
      </Wrapper>
    );
  }

  const handleSubmit = async (e) => {
    e.preventDefault();
    e.currentTarget.disabled = true;
    const response = await fetchAsync(`/levels/${levelNum}/leaderboard`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ name }),
    });

    setUpdatedData(response);
  };

  const getMenuBody = () => {
    // Score not submitted yet
    if (updatedData === null) {
      return (
        <>
          <Score>
            Your Score - <b>{formatTime(score)}</b>
          </Score>
          <Leaderboard data={data} loading={loading} />
          <form>
            <label htmlFor="name">Name</label>
            <Input
              type="text"
              id="name"
              maxLength={32}
              value={name}
              onChange={(e) => setName(e.target.value)}
            />
            <Button type="submit" onClick={handleSubmit}>
              Submit
            </Button>
          </form>
        </>
      );
    }

    // Error submitting score
    if (updatedData instanceof Error) {
      return (
        <>
          <p>There was an error contacting the server.</p>
          <Button type="button" onClick={() => navigate("/")}>
            Home
          </Button>
        </>
      );
    }

    // Score submitted
    return (
      <>
        <Leaderboard data={updatedData} loading={loading} />
        <Button onClick={() => navigate("/")}>Finish</Button>
      </>
    );
  };

  return (
    <Wrapper>
      <Menu>
        <Heading>LEVEL COMPLETE!</Heading>
        {getMenuBody()}
      </Menu>
    </Wrapper>
  );
};

LevelCompleteMenu.propTypes = {
  levelNum: PropTypes.string.isRequired,
  score: PropTypes.number.isRequired,
  token: PropTypes.string.isRequired,
};

export default LevelCompleteMenu;
