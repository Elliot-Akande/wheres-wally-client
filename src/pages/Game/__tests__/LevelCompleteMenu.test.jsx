import { render, screen } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import LevelCompleteMenu from "../LevelCompleteMenu/LevelCompleteMenu";
import { MemoryRouter } from "react-router-dom";
import { userEvent } from "@testing-library/user-event";
import fetchAsync from "../../../utils/fetchAsync";

const mocks = vi.hoisted(() => ({
  useFetch: vi.fn(),
  fetchAsync: vi.fn(),
}));

vi.mock("../../../hooks/useFetch.jsx", () => ({
  default: mocks.useFetch,
}));
vi.mock("../../../utils/fetchAsync.js", () => ({
  default: mocks.fetchAsync,
}));

vi.mock("../../../components/Leaderboard/Leaderboard.jsx", () => ({
  default: () => <div>Leaderboard</div>,
}));

const levelNum = 1;
const score = 100;
const token = "token";
const data = [
  {
    _id: "123",
    name: "john",
    score: 30,
  },
  {
    _id: "178",
    name: "bob",
    score: 120,
  },
];

describe("LevelCompleteMenu", () => {
  it("renders title, score, leaderboard, name form, and submit button", () => {
    mocks.useFetch.mockReturnValue({ data, loading: false, error: null });

    const { container } = render(
      <MemoryRouter>
        <LevelCompleteMenu levelNum={levelNum} score={score} token={token} />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("renders message on initial data fetching error", () => {
    mocks.useFetch.mockReturnValue({
      data: null,
      loading: false,
      error: new Error(),
    });

    const { container } = render(
      <MemoryRouter>
        <LevelCompleteMenu levelNum={levelNum} score={score} token={token} />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("renders title, leaderboard, and finish button on score submission", async () => {
    const user = userEvent.setup();
    const updatedData = [
      {
        _id: "123",
        name: "john",
        score: 30,
      },
      {
        _id: "369",
        name: "tim",
        score: 100,
      },
      {
        _id: "178",
        name: "bob",
        score: 120,
      },
    ];
    mocks.useFetch.mockReturnValue({
      data,
      loading: false,
      error: null,
    });
    mocks.fetchAsync.mockResolvedValue(updatedData);

    const { container } = render(
      <MemoryRouter>
        <LevelCompleteMenu levelNum={levelNum} score={score} token={token} />
      </MemoryRouter>
    );
    await user.clear(screen.getByRole("textbox", { name: "Name" }));
    await user.type(screen.getByRole("textbox", { name: "Name" }), "tim");
    await user.click(screen.getByRole("button", { name: "Submit" }));

    const body = JSON.parse(fetchAsync.mock.lastCall[1].body);
    expect(body.name).toBe("tim");
    expect(container).toMatchSnapshot();
  });

  it("renders message on updated data fetching error", async () => {
    const user = userEvent.setup();
    mocks.useFetch.mockReturnValue({
      data,
      loading: false,
      error: null,
    });
    mocks.fetchAsync.mockResolvedValue(new Error());

    const { container } = render(
      <MemoryRouter>
        <LevelCompleteMenu levelNum={levelNum} score={score} token={token} />
      </MemoryRouter>
    );
    await user.click(screen.getByRole("button", { name: "Submit" }));

    expect(container).toMatchSnapshot();
  });
});
