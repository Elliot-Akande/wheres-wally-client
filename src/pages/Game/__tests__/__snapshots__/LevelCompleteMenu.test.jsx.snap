// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LevelCompleteMenu > renders message on initial data fetching error 1`] = `
<div>
  <div
    class="_container_a7af52"
  >
    <div
      class="_menu_a7af52"
    >
      <h2>
        Level complete!
      </h2>
      <p>
        There was an error contacting the server.
      </p>
      <button
        type="button"
      >
        Home
      </button>
    </div>
  </div>
</div>
`;

exports[`LevelCompleteMenu > renders message on updated data fetching error 1`] = `
<div>
  <div
    class="_container_a7af52"
  >
    <div
      class="_menu_a7af52"
    >
      <h1
        class="_heading_a7af52"
      >
        LEVEL COMPLETE!
      </h1>
      <p>
        There was an error contacting the server.
      </p>
      <button
        class="_button_a7af52"
        type="button"
      >
        Home
      </button>
    </div>
  </div>
</div>
`;

exports[`LevelCompleteMenu > renders title, leaderboard, and finish button on score submission 1`] = `
<div>
  <div
    class="_container_a7af52"
  >
    <div
      class="_menu_a7af52"
    >
      <h1
        class="_heading_a7af52"
      >
        LEVEL COMPLETE!
      </h1>
      <div>
        Leaderboard
      </div>
      <button
        class="_button_a7af52"
      >
        Finish
      </button>
    </div>
  </div>
</div>
`;

exports[`LevelCompleteMenu > renders title, score, leaderboard, name form, and submit button 1`] = `
<div>
  <div
    class="_container_a7af52"
  >
    <div
      class="_menu_a7af52"
    >
      <h1
        class="_heading_a7af52"
      >
        LEVEL COMPLETE!
      </h1>
      <p
        class="_score_a7af52"
      >
        Your Score - 
        <b>
          01:40
        </b>
      </p>
      <div>
        Leaderboard
      </div>
      <form
        class="_form_a7af52"
      >
        <label
          class="_label_a7af52"
          for="name"
        >
          Name
        </label>
        <input
          class="_input_a7af52"
          id="name"
          maxlength="32"
          type="text"
          value="Anon"
        />
        <button
          class="_button_a7af52"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</div>
`;
